### PR TITLE
Add linting and pre-commit hooks

### PR DESCRIPTION
**Pull request by jellis18**
This MR implements pre-commit hooks that run `black`, `isort`, and `flake8`. At the moment these are installed via the `make init` command but I don't know how many developers use that since I know most use `conda`. There can be a different command if needed.

All the files are formatted now and all files follow flake8. There are no code changes other than removing unneeded imports except for a few bugs I found from funning flake8. I'll comment on those below.

### CODE REVIEW

File: enterprise_extensions/deterministic.py [hunk 1]
```diff
@@ -438,7 +467,7 @@ def bwm_delay(toas, pos, log10_h=-14.0, cos_gwtheta=0.0, gwphi=0.0, gwpol=0.0, t
 
     # antenna patterns
     if antenna_pattern_fn is None:
-        apc = create_gw_antenna_pattern(pos, gwtheta, gwphi)
+        apc = utils.create_gw_antenna_pattern(pos, gwtheta, gwphi)
```
Comment (jellis18) [inline enterprise_extensions/deterministic.py:470]: This was a bug spotted  by `flake8`

File: enterprise_extensions/models.py [hunk 2]
```diff
@@ -2471,47 +2839,66 @@ def model_bwm_sglpsr (psr, likelihood=LogLikelihood, lookupdir=None, noisedict=N
     # adding white-noise, and acting on psr objects
     models = []
 
-    if 'NANOGrav' in psr.flags['pta'] and not wideband:
+    if "NANOGrav" in psr.flags["pta"] and not wideband:
         s2 = s + white_noise_block(vary=False, inc_ecorr=True)
-        if dm_var and 'J1713+0747' == p.name:
+        if dm_var and "J1713+0747" == psr.name:
```
Comment (jellis18) [inline enterprise_extensions/models.py:2844]: This was also a bug found by flake8

**Comment by paulthebaker** [comment]
I'm all for linting and isort.  I think `black` can make some code harder to read, particularly complicated equations.  There are many places where we want the linter to ignore "space between operator" errors.  For instance, this line from the `cw_delay` in `deterministic.py`:

```python
fstart = fgw * (1 - 256/5 * mc**(5/3) * fgw**(8/3) * toas[0])**(-3/8)  # original

fstart = fgw * (1 - 256 / 5 * mc ** (5 / 3) * fgw ** (8 / 3) * toas[0]) ** (-3 / 8)  # new
```
This is a pretty minor change, but it's way easier for me to read without all the space.  Some terms go together, and the lack of space emphasizes that point.

I would be in favor of using something like `black`, if would configure which rules we want to enforce.

`black` would save us the arduous work of having to manually fix a whole lot that `flake8` will complain about.  To that end, there's an argument to be made that we should "blacken" the code now, but I don't think we want to enforce the black style on all future code.  Even if we do "blacken" the code one time only, we might want to manually undo some reformating

**Comment by kdolum** [comment]
Vertical spacing is a problem in `black` also, in the same way that @paulthebaker pointed out for horizontal.  You might want to group some code by writing it without blank lines and set it off from some other code by leaving blank lines.  In general I think the "any color you want as long as it's black" philosophy denies us the ability to format code ourselves to increase readability.

Also in my opinion the actual format used by `black` is lousy.  If we wanted to have a standard format, this is not the one I'd use.  But I guess there's no arguing about taste.

**Comment by jellis18** [comment]
You can wrap a code block in `#fmt: off`, `#fmt: on` if you want black to ignore it but in my opinion you are in for a world of pain if you don't use black and you use flake8. But I don't have to deal with this code so its up to you guys

**Comment by kdolum** [comment]
Maybe we could not use any tool that enforces style conventions but only one which finds actual errors such as undefined functions.

**Comment by Hazboun6** [comment]
How are we going to resolve this? Obviously we need linting. But, I also agree with @jellis18 that this code is a bit of a mess. 

We are about to undertake a code review, so there is benefit to having the code be easily readable by the reviewers. I definitely don't have the time to go through the code and surround specific code blocks with `#fmt:#

BTW `black` not being the best format option is not a new opinion:
image

**Comment by Hazboun6** [comment]
@stevertaylor and @svigeland as writers of a good portion of this code what do you think?

**Comment by Hazboun6** [comment]
Okay I think I have a compromise. I'll work on porting over all of the helpful things @jellis18 has included minus `black` and I'm going to run the code through [autopep8](https://github.com/hhatto/autopep8). I also used `autoflake` and `isort`, which were definitely not as easy as `black`, but much easier than doing things by hand. I kept the changes pretty light, but there are now no pep8 errors. Will submit a different PR if people think this is a good idea.

**Comment by Hazboun6** [comment]
Okay thanks @jellis18 for pointing out the mistakes that come from a lack of linting checks! I've submitted a different PR #140 that reflects the dev's desires for readability over the use of `black`.

**Pull request closed without merging**